text/0000-pack-package-refactor.md

```diff
@@ -0,0 +1,118 @@
+
+# What it is
+[what-it-is]: #what-it-is
+The general flow of this process is
+1) Unorganized features exist & are added to the `internal` package, these are not consumable by external programs
+2) Extract code from the `interal` package to `/pkg/<pkg-name>/...` to enable external consumption.
+3) Provide some useful set of functionality through independent binaries at `/cmd/<bin-name>` that other platforms may consume.
+4) Extract code from external packages `/pkg/<pkg-name>/...` into a separate repository (this should require an RFC)
+
+The goal of this RFC is to:
+- Reach consensus on the overall code promotion process for the `pack` repo
+- Reach consensus on a "probable" target structure for `pack`
+
+The goal of this RFC is NOT:
+- to precisely define what packages we want & what their contents/functionality should be.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+### Current structure
+The current state of the `pack` repo does not provide a standardized structure, and much of the logic
+in the inaccessible `internal` package.
+```
+.
+├── builder
+├── buildpackage
+├── cmd
+│    └── pack
+├── config
+├── internal/...
+├── logging
+├── out
+│    └── tests
+├── pkg
+│    └── archive
+├── project
+├── registry
+├── resources
+└── tools
+    └── pedantic_imports
+```
+
+### Short term target structure
+This is a short term approximation for the structure that we would like to see. This is not set in stone
+and may be updated during implementation. 
+
+```
+.
+├── cmd/
+│   └── pack/ → current pack cli
+├── pkg/
+│   ├── archive/ → lib for converting multiple sources into OCI layers
+│   ├── client/ → current pack client library
+│   ├── descriptor/ → lib for processing project descriptor
+│   ├── inspect/ → lib for inspecting various CNB resources (app image, buildpack, builder, stack, ...)
+│   ├── internal/ → all non-exported go code
+│   └── resolve/ → lib for resolving buildpacks via Buildpacks URIs
+└── go.mod
+```
+
+### Possible Future structure.
+This is a long term guess about which functionality we would want to pull out into packages as well as additional 
+commands that could be used by other platforms (such as Tekton)
+
+```
+.
+├── cmd/
+│   ├── pack/ → current pack cli
+│   ├── env-vars/ → cli for handling platform env vars.
+│   └── fetcher/ → cli to download buildpacks.
+├── pkg/
+│   ├── client/ → current pack client library
+│   ├── descriptor/ → lib for processing project descriptor
+│   ├── inspect/ → lib for inspecting various CNB resources (app image, buildpack, builder, stack, ...)
+│   ├── internal/ → all non-exported go code
+│   ├── stack/ → lib for mixin validation & stack labels
+│   ├── builder/ → lib for builder labels & builder manipulation
+│   ├── buildpackage/ → lib for buildpackage label & buildpack extraction from buildpackages
+│   └── resolve/ → lib for resolving buildpacks via Buildpacks URIs
+└── go.mod
+
+archive → pulled to a separate repo.
```

Comment:
Assumedly part of `imgutil`, per the discussion here: https://github.com/buildpacks/imgutil/pull/73